packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts

```diff
@@ -3,75 +3,102 @@
-      existingRow[field] = value;
-    }
+const sum = (values: FieldValue[]): number | undefined => {
+  const checkedValues = checkIsNum(values);
+  return checkedValues.length === 0
+    ? undefined
+    : checkedValues.reduce((a, b) => {
+        return a + b;
+      });
+};
+
+const avg = (values: FieldValue[]): number | undefined => {
+  const checkedValues = checkIsNum(values);
+  const numerator = sum(checkedValues);
+  const denominator = count(checkedValues);
+  if (isNotUndefined(numerator) && denominator !== 0) {
+    return numerator / denominator;
   }
+  return undefined;
+};
+
+const count = (values: FieldValue[]): number => {
+  return values.length;
 };
 
-const min = (existingRow: Row, field: string, value: FieldValue) => {
-  const existingValue: FieldValue = existingRow[field];
-  if (!isUndefined(value)) {
-    if (isUndefined(existingValue)) {
-      existingRow[field] = value;
-    } else if (value < existingValue) {
-      existingRow[field] = value;
+// helper of max() and min(), e.g.: customSort([1, 2, 3, null]) = [null, 1, 2, 3]
+const customSort = (values: FieldValue[]): FieldValue[] => {
+  const filteredUndefinedValues = values.filter(isNotUndefined);
+  return filteredUndefinedValues.sort((a, b) => {
+    if (a === null) {
+      return -1;
     }
-  }
+    if (b === null) {
+      return 1;
+    }
+    return a > b ? 1 : -1;
+  });
 };
 
-const unique = (existingRow: Row, field: string, value: FieldValue) => {
-  if (!isUndefined(existingRow[field]) && existingRow[field] !== value) {
-    existingRow[field] = 'NO_UNIQUE_VALUE';
-  } else {
-    existingRow[field] = value;
-  }
+// max([1, 2, 3, null]) = 3; max([null, null]) = null
+const max = (values: FieldValue[]): FieldValue => {
+  const sortedValues = customSort(values);
+  return sortedValues.length !== 0 ? sortedValues.reverse()[0] : undefined;
+};
+
+// min([1, 2, 3, null]) = null; min([null, null]) = null
+const min = (values: FieldValue[]): FieldValue => {
+  const sortedValues = customSort(values);
+  return sortedValues.length !== 0 ? sortedValues[0] : undefined;
 };
 
-const drop = (existingRow: Row, field: string, value: FieldValue) => {
+const unique = (values: FieldValue[]): FieldValue => {
+  const distinctValues = [...new Set(values.filter(isNotUndefined))];
+  return distinctValues.length === 1 ? distinctValues[0] : 'NO_UNIQUE_VALUE';
+};
+
+const drop = (values: FieldValue[]): FieldValue => {
   // Do nothing, don't add the field to the existing row
 };
 
-const first = (existingRow: Row, field: string, value: FieldValue) => {
-  if (isUndefined(existingRow[field])) {
-    existingRow[field] = value;
-  }
+const first = (values: FieldValue[]): FieldValue => {
+  return values.find(isNotUndefined);
 };
 
-const last = (existingRow: Row, field: string, value: FieldValue) => {
-  existingRow[field] = value;
+const last = (values: FieldValue[]): FieldValue => {
+  return values.reverse().find(isNotUndefined);
```

Comment:
You gotta be a bit careful about calling `reverse` on an array, as it actually changes the data in the base array (as opposed to something like `map` that creates a copy of the base array with the values mapped)

Usually people do something like this `array.slice().reverse()` as `slice()` creates a copy of the array 👍

Reply:
Sorry to miss one of your reviews that using `slice()`.. Thank you to point it out again and share the good practice!